File: Striver A2Z/Step 6 - Learn LinkedList [Single LL, Double LL, Medium, Hard Problems] /Lec 3 - Medium Problems of LL/remove-nth-node-from-end-of-list.cpp

// https://leetcode.com/problems/remove-nth-node-from-end-of-list/description/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:

    ListNode*removeNthFromEnd(ListNode*head, int n, int &i){
        if(!head) return head;
        head->next = removeNthFromEnd(head->next, n, i);
        i+=1;
        if(i == n) return head->next;
        return head;
    }

    ListNode* removeNthFromEnd(ListNode* head, int n) {
        if(!head) return head;
        int i = 0;
        return removeNthFromEnd(head, n, i);
    }
};

File: Neetcode-150-js/Linked List/remove-nth-node-from-end-of-list.js

// https://leetcode.com/problems/remove-nth-node-from-end-of-list/description/

/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */

let i = 0;

var helper = function(head, n){
    if(!head) return head;
    head.next  = removeNthFromEnd(head.next, n);
    i++;
    if(i === n) return head.next;
    return head;
}

var removeNthFromEnd = function(head, n) {
    if(!head) return head;
    i = 0;
    return helper(head, n);
};

